cumulusci/tasks/bulkdata/delete.py

```diff
@@ -38,77 +41,46 @@ def _init_options(self, kwargs):
                 "Criteria cannot be specified if more than one object is specified."
             )
         self.options["hardDelete"] = process_bool_arg(self.options.get("hardDelete"))
+        self.options["ignore_row_errors"] = process_bool_arg(
+            self.options.get("ignore_row_errors")
+        )
 
     def _run_task(self):
         for obj in self.options["objects"]:
             self.logger.info(f"Deleting {self._object_description(obj)} ")
-            delete_job = self._create_job(obj, self.options["where"])
-            if delete_job is not None:
-                self._wait_for_job(delete_job, error_behaviour="raise")
-
-    def _create_job(self, obj, where=None):
-        # Query for rows to delete
-        delete_rows = self._query_salesforce_for_records_to_delete(obj, where)
-        if not delete_rows:
-            self.logger.info(f"  No {obj} objects found, skipping delete")
-            return
 
-        # Upload all the batches
-        operation = "hardDelete" if self.options["hardDelete"] else "delete"
-        delete_job = self.bulk.create_job(obj, operation)
-        self.logger.info(f"  Deleting {len(delete_rows)} {obj} records")
-        batch_num = 1
-        for batch in self._upload_batches(delete_job, delete_rows):
-            self.logger.info(f"    Uploaded batch {batch}")
-            batch_num += 1
-        self.bulk.close_job(delete_job)
-        return delete_job
-
-    def compose_query(self, obj, where):
-        query = f"SELECT Id FROM {obj}"
-        if where:
-            query += f" WHERE {where}"
+            query = f"SELECT Id FROM {obj}"
+            if self.options["where"]:
+                query += f" WHERE {self.options['where']}"
+
+            if not self.options["where"]:
+                # FIXME: Perform a count() query to determine whether we have work to do.
+                pass
+
+            qs = BulkApiQueryStep(obj, {}, self, query)
```

Comment:
Would it be virtuous and helpful to have these be Tasks? We should be conservative about adding new abstractions (but of course should do so when it is helpful).

Reply:
My design goal was to isolate the API access layer to act as sort of Bulk API primitives that are easily composable. It's expected that individual tasks will use multiple `Step` instances in many use cases - for example, `DeleteData` uses both a `QueryStep` and a `DmlStep`. for that reason and because the steps do not have a concept of a storage layer, I think they belong as a separate abstraction, but I'm open to pushback and discussion on that.

Reply:
I like the isolation and code reuse. I'm in favour of that. I'm suggesting that the coupling be even looser than it is, rather than tighter.

I'm suggesting roughly that instead of

            qs = BulkApiQueryStep(obj, {}, self, query)

It would delegate to a subtask more like this:

            BulkApiQueryTask({output_file=my_csv_tempfile_path, query=query})

...

           BulkApiDmltask({operation = Operation.HARD_DELETE, id_csv=my_csv_tempfile_path})

Although the actual syntax might not be so clean unless we make a special helper for spawning sub-tasks. Which we probably could and should. I could have used it here:

https://github.com/SFDO-Tooling/CumulusCI/blob/4cf572a6703b7c3244bfd7514ad2efc24b1e5162/cumulusci/tasks/bulkdata/generate_and_load_data.py#L144

That would mean, among other things, that I could generate my own QueryResults file on the command line (using whatever complicated Python code or spreadsheet hackery I want) and then I could also pass it to a DmlStep to do the actual deleting (or Loading, or Updating). This would give CCI a user-accessible mechanism for deleting a list of IDs (wherever they came from) and *on top of that* a mechanism for deleting from a query, if the list of objects I want to delete can be expressed easily that way. (just this week I ran into a case where I had to write custom Python code to delete a list of objects that was not easily expressible as a simple query result).

Another thing it would mean is that the end user could try and redo a DmlTask if it fails (as long as they have a way of generating the list of records to delete, insert or update).

Reply:
I would prefer to build that functionality as a Task layer on top of these classes. As implemented here, they know nothing about input and output - they just consume a generator and produce another generator of results. I'd like to separate the logic that relates to storage and persistence.

Reply:
Okay that works too.